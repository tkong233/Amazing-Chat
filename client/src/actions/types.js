// auth
export const GET_ERRORS = "GET_ERRORS";
export const USER_LOADING = "USER_LOADING";
export const SET_CURRENT_USER = "SET_CURRENT_USER";
export const UPDATE_PICTURE = "UPDATE_PICTURE";
export const DELETE_ACCOUNT = "DELETE_ACCOUNT";
export const SET_SOCKET = "SET_SOCKET";
export const UNSET_SOCKET = "UNSET_SOCKET";

// contact
export const GET_SUGGESTION = "GET_SUGGESTION";
export const GET_CONTACTS = "GET_CONTACTS";
export const ADD_CONTACT = "ADD_CONTACT";
export const DELETE_CONTACT = "DELETE_CONTACT";
export const GET_ALL_USERS = "GET_ALL_USERS";

// status
export const GET_STATUS = "GET_STATUS";

// chat
export const CONNECT_SOCKET = "CONNECT_SOCKET";
export const DISCONNECT_SOCKET = "DISCONNECT_SOCKET";
export const JOIN_ROOM = "JOIN_ROOM";
export const LOAD_PAST_MESSAGES = "LOAD_PAST_MESSAGES";
export const RECEIVE_NEW_MESSAGES = "RECEIVE_NEW_MESSAGES";
export const SEND_MESSAGE = "SEND_MESSAGE";
export const SET_ITEM_STATUS = "SET_ITEM_STATUS";
export const DELETE_MESSAGE = "DELETE_MESSAGE";
export const INITIATE_VIDEO_CALL = "INITIATE_VIDEO_CALL";
export const RECEIVE_VIDEO_CALL = "RECEIVE_VIDEO_CALL";
export const PICK_UP_VIDEO_CALL = "PICK_UP_VIDEO_CALL";
export const HANG_UP_VIDEO_CALL = "HANG_UP_VIDEO_CALL";
export const SET_CALLEE_ONLINE = "SET_CALLEE_ONLINE";
export const SET_CALLEE_OFFLINE = "SET_CALLEE_OFFLINE";
export const STOP_WAITING_FOR_CALLEE_RESPONSE = "STOP_WAITING_FOR_CALLEE_RESPONSE";