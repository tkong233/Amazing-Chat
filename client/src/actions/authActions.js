import axios from "axios";
import io from 'socket.io-client';
import setAuthToken from "../utils/setAuthToken";
import jwt_decode from "jwt-decode";

import {
  SET_SOCKET,
  UNSET_SOCKET,
  GET_ERRORS,
  SET_CURRENT_USER,
  USER_LOADING,
  UPDATE_PICTURE,
  DELETE_ACCOUNT
} from "./types";

// Register User
export const registerUser = (userData, history) => dispatch => {
  axios
    .post("/api/users/register", userData)
    .then(res => history.push({    
        pathname: "/login",
        state: true}))
    .catch(err => {
      // console.log(err.response.data);
      dispatch({
        type: GET_ERRORS,
        payload: err.response.data
      })}
    );
};

// Login - get user token
export const loginUser = userData => dispatch => {
  axios
    .post("/api/users/login", userData)
    .then(res => {
      // Save to localStorage

      // Set token to localStorage
      const { token } = res.data;
      localStorage.setItem("jwtToken", token);
      // Set token to Auth header
      setAuthToken(token);
      // Decode token to get user data
      const decoded = jwt_decode(token);
      // Set current user
      dispatch(setCurrentUser(decoded));
      // create socket and store in reducer state
      const socket = io();
      socket.emit('sendEmail', { email: userData.email })
      dispatch({
        type: SET_SOCKET,
        payload: socket,
      })
    })
    .catch(err =>
      dispatch({
        type: GET_ERRORS,
        payload: err.response.data
      })
    );
};

// Set logged in user
export const setCurrentUser = decoded => {
  return {
    type: SET_CURRENT_USER,
    payload: decoded
  };
};

export const setSocket = (email) => {
  const socket = io();
  socket.emit('email', { email });
  return {
    type: SET_SOCKET,
    payload: socket,
  }
}

// User loading
export const setUserLoading = () => {
  return {
    type: USER_LOADING
  };
};

// Log user out
export const logoutUser = (socket) => dispatch => {
  // Remove token from local storage
  localStorage.removeItem("jwtToken");
  // Remove auth header for future requests
  setAuthToken(false);
  // Set current user to empty object {} which will set isAuthenticated to false
  dispatch(setCurrentUser({}));
  socket.disconnect();
  dispatch({
    type: UNSET_SOCKET
  })
};

export const resetPassword = (userData, history) => dispatch => {
    axios
    .post("/api/users/reset", userData)
    .then(res => history.push("/login"))
    .catch(err =>
      dispatch({
        type: GET_ERRORS,
        payload: err.response.data
      })
    );
}

export const updatePicture = (formData, email) => dispatch => {
  axios.post
  (`/api/users/upload_profile_image/${email}`, formData, {
    headers: {
      "Content-Type": "multipart/form-data",
    },
  }).then(res => {
    dispatch({
      type: UPDATE_PICTURE,
      payload: res.data.user
    })
  }).catch(err => console.log(err));
}

export const deleteAccount = (email) => dispatch => {
  axios
  .delete(`/api/users/profile/${email}`)
  .then(res => dispatch({
    type: DELETE_ACCOUNT, 
  }))
  .catch(err => console.log(err))
  
}

export const doPostStatus = (formData, email, history) => dispatch => {
  axios.post(`/api/users/status/${email}`, formData, {
    headers: {
      "Content-Type": "multipart/form-data",
    },
  }).then(res => history.push({    
    pathname: "/dashboard",
    state: true}))
  .catch(err => console.log(err))
}