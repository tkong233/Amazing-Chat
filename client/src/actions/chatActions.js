import axios from "axios";
import {
  CONNECT_SOCKET,
  JOIN_ROOM,
  LOAD_PAST_MESSAGES,
  RECEIVE_NEW_MESSAGES,
  SEND_MESSAGE,
  DISCONNECT_SOCKET,
  RECEIVE_VIDEO_CALL,
  PICK_UP_VIDEO_CALL,
  HANG_UP_VIDEO_CALL,
  INITIATE_VIDEO_CALL,
} from "./types";

export const connectSocket = (socket) => (dispatch) => {
  dispatch({
    type: CONNECT_SOCKET,
    payload: socket,
  });
};

export const joinRoom = (
  senderName,
  receiverName,
  sender,
  receiver,
  pairId,
  socket
) => (dispatch) => {
  const data = {
    name: senderName,
    pairId,
  };

  if (socket) {
    socket.emit("join", data);
    dispatch({
      type: JOIN_ROOM,
      payload: { pairId, sender, receiver, senderName, receiverName },
    });
  } else {
    console.log("can not join room because socket is not valid");
  }
};

export const loadPastMessages = (pairId) => (dispatch) => {
  axios
    .get(`/messages/${pairId}`)
    .then((res) => {
      dispatch({
        type: LOAD_PAST_MESSAGES,
        payload: res.data,
      });
    })
    .catch((err) => {
      console.log(err);
    });
};

export const receiveNewMessages = (socket) => (dispatch) => {
  socket.on("receiveMessage", (message) => {
    console.log("receive message", message);
    dispatch({
      type: RECEIVE_NEW_MESSAGES,
      payload: message,
    });
  });
};

export const sendMessage = (data, socket) => (dispatch) => {
  const { pairId, from, to, message, type } = data;
  console.log(message);

  axios
    .post("/message", { pairId, from, to, message, type })
    .then((res) => {
      const { datetime } = res.data.savedMessage;
      socket.emit(
        "sendMessage",
        { pairId, from, to, message, datetime, type },
        message
      );
      dispatch({
        type: SEND_MESSAGE,
        payload: res.data.savedMessage,
      });
    })
    .catch((err) => {
      console.log(err);
    });
};

export const initiateVideoCall = (pairId, from, to, socket) => (dispatch) => {
  console.log("action: initiate video call");
  socket.emit("initiateVideoCall", { pairId, from, to });
  dispatch({
    type: INITIATE_VIDEO_CALL,
  });
};

export const receiveVideoCall = (socket, email) => (dispatch) => {
  console.log("action: receive vedio call");
  socket.on("receiveVideoCall", ({ pairId, from, to }) => {
    console.log("socket: receive video call", pairId, from, to);
    if (email === to) {
      console.log("I am video call receiver");
      dispatch({
        type: RECEIVE_VIDEO_CALL,
      });
    } else {
      console.log("I am video call sender");
    }
  });
};

export const pickUpVideoCall = (socket, pairId, from, to) => (dispatch) => {
  console.log("action: pick up video call");
  socket.emit("videoCallRequestDecision", { pairId, from, to, accept: true });
  dispatch({
    type: PICK_UP_VIDEO_CALL,
  });
};

export const hangUpVideoCall = (socket, pairId, from, to) => (
  dispatch
) => {
  console.log("action: hang up video call");
  

  socket.emit("videoCallRequestDecision", { pairId, from, to, accept: false });

  dispatch({
    type: HANG_UP_VIDEO_CALL,
  });
};

export const waitForVideoCallDecision = (socket) => (dispatch) => {
  socket.on("videoCallRequestResult", ({ accept }) => {
    if (accept) {
      console.log("socket: video call request accepted");
      dispatch({
        type: PICK_UP_VIDEO_CALL,
      });
    } else {
      console.log("socket: video call request rejected");
      dispatch({
        type: HANG_UP_VIDEO_CALL,
      });
    }
  });
};

export const uploadChatFiles = (formData, socket) => (dispatch) => {
  axios
    .post("/message/uploadfiles", formData, {
      headers: {
        "content-type": "multipart/form-data",
      },
    })
    .then((res) => {
      const {
        pairId,
        from,
        to,
        message,
        datetime,
        type,
      } = res.data.savedMessage;
      socket.emit(
        "sendMessage",
        { pairId, from, to, message, datetime, type },
        message
      );
      dispatch({
        type: SEND_MESSAGE,
        payload: res.data.savedMessage,
      });
    });
};

export const disconnectSocket = (socket) => (dispatch) => {
  if (socket) {
    socket.disconnect();
  } else {
    console.log("socket is not valid, can not disconnect");
  }

  dispatch({
    type: DISCONNECT_SOCKET,
  });
};
