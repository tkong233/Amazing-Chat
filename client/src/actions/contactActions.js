import axios from "axios";

import { GET_CONTACTS, GET_SUGGESTION, ADD_CONTACT, DELETE_CONTACT } from "./types";

// Get suggestion
export const getSuggestion = (email) => dispatch => {
  axios
    .get(`/suggestion/${email}`)
    .then(res => {
      dispatch({
        type: GET_SUGGESTION,
        payload: res.data
      })
    })
    .catch(err => {
      console.log(err);
    })
}

export const addContact = (email1, email2) => dispatch => {
  axios
    .post('/contact', { email1, email2 })
    .then(res => {
      console.log(res.data);
      dispatch({
        type: ADD_CONTACT,
        payload: res.data
      })
    })
    .catch(err => {
      console.log(err);
    })
}
  
export const deleteContact = (email1, email2) => dispatch => {
  axios
  .delete('/contact', { email1, email2 })
  .then(res => {
    console.log(res.data);
    dispatch({
      type: DELETE_CONTACT,
      payload: res.data
    })
  })
  .catch(err => {
    console.log(err);
  })
}

export const getContacts = (email) => dispatch => {
  axios
    .get(`/contacts/${email}`)
    .then(res => {
      dispatch({
        type: GET_CONTACTS,
        payload: res.data,
      })
    })
    .catch(err => {
      console.log(err);
    })
}