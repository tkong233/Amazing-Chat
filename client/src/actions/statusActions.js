import axios from "axios";

import { GET_, GET_SUGGESTION, ADD_CONTACT, DELETE_CONTACT } from "./types";