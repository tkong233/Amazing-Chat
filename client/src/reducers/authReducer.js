import { SET_CURRENT_USER, USER_LOADING, UPDATE_PICTURE, DELETE_ACCOUNT } from "../actions/types";

const isEmpty = require("is-empty");

const initialState = {
  isAuthenticated: false,
  user: {},
  loading: false
};

export default function(state = initialState, action) {
  switch (action.type) {
    case SET_CURRENT_USER:
      return {
        ...state,
        isAuthenticated: !isEmpty(action.payload),
        user: action.payload
      };
    case USER_LOADING:
      return {
        ...state,
        loading: true
      };
    case UPDATE_PICTURE:
      return {
        ...state,
        user: action.payload
      }
    case DELETE_ACCOUNT:
      return {
        ...state,
        user: {}
      }
    default:
      return state;
  }
}
