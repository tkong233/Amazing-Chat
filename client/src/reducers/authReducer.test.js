import { SET_CURRENT_USER, USER_LOADING, UPDATE_PICTURE, DELETE_ACCOUNT } from "../actions/types";
import authReducer from './authReducer';

describe('Auth Reducer', ()=>{
    it('Should return default state', ()=>{
        const newState = authReducer(undefined, {});
        expect(newState).toEqual({
            isAuthenticated: false,
            user: {},
            loading: false});
    });
    it ('Should return new state if receiving type', () =>{
        const user = {isAuthenticated: false,
            user: {},
            loading: true};
        const newState = authReducer(undefined, {
            type: USER_LOADING,
            payload: user
        });
        expect(newState).toEqual(user);
    });
    it ('Should delete account', () =>{
        const user = {isAuthenticated: false,
            user: {},
            loading: false};
        const newState = authReducer(undefined, {
            type: DELETE_ACCOUNT,
            payload: user
        });
        expect(newState).toEqual(user);
    });
    it ('should return initial state', ()=>{
        expect(authReducer(undefined, {})).toMatchSnapshot();
    });
});