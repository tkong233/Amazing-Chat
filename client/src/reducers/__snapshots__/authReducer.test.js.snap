// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Reducer should return initial state 1`] = `
Object {
  "isAuthenticated": false,
  "loading": false,
  "user": Object {},
}
`;
